IshiHisashi/GlucoFit
# feat : JDEV-103 | Migrate badge rewards logic into createTestResult

## Summary
- This is to enable the app to reward when a certain number of streak (testResult) is achieved

## Changes
- `createTestResult` in `testResultResolvers.tsx` : The function name was changed from former `createTestResultWithInsight` as this should work for multiple features. After insight section, the logic is added. It will check rewarding of 7 badges that are related to testResult streak.
- The return value of above function is now `return { articlesToShow, badgesToShow }`.
- Since streak calc appears many times, I created another file called `testResultUtil.ts` and `testResultResolvers.tsx` imports the utilities for streak calc. 

## Core logic in createTestResult
- Firstly, it will check which badges are 'unachieved' for an user. 
- Based on those unachieved badgeID, streak calcs are ran and the result matchs with each of criteria. If it's met, the corresponding badge is rewarded. 
- The reward is logged on User doc (badges array) by turning to `achieved : true`.
- And, like insight, corresponding badge doc is stored in the badgesToShow array (of badge doc), for further notification on the home screen.

## Notes
- The rewarding logic still have some repeating section. This is a room for furthre refactoring but I haven't made it.

## How to test
- Examine with userId `670db268582e7e887e447288`. This guy has 8 streak already but the badge 'Streak Starter' and 'Healthy habit' have not been given. Now the new streak should reward those badges as this guy had already achieved the criteria, actually.
- Please run below on the appolo, then you will see two new badges are rewarded. If you execute the same function again, no badges are contained in the return as they were already rewarded. 
```
mutation CreateTestResult {
  createTestResult(
    user_id: "670db268582e7e887e447288",
    bsl: 120.5,
    time_period: "morning",
    confirmed: true
  ) {
    articlesToShow {
      article_name
    }
    badgesToShow {
      badge_name
      badge_desc
    }
  }
}
```
<img width="565" alt="image" src="https://github.com/user-attachments/assets/628a5297-17a6-4d46-bbad-84b3f16c5144">
